File: Python/0123-h-best-time-to-buy-and-sell-stock-iii.py
class Solution_1:
    def maxProfit(self, prices: List[int]) -> int:
        hold1, hold2 = float('-inf'), float('-inf')
        release1, release2 = 0, 0
        for i in prices:
            # The maximum if we've just sold 2nd stock so far.
            release2 = max(release2, hold2 + i)
            # The maximum if we've just buy  2nd stock so far.
            hold2 = max(hold2, release1 - i)
            # The maximum if we've just sold 1nd stock so far.
            release1 = max(release1, hold1 + i)
            # The maximum if we've just buy  1st stock so far. 
            hold1 = max(hold1, -i)
        #Since release1 is initiated as 0, so release2 will always higher than release1.
        return release2

"""
f[k, ii] represents the max profit up until prices[ii] (Note: NOT ending with prices[ii]) using at most k transactions. 
f[k, ii] = max(f[k, ii-1], prices[ii] - prices[jj] + f[k-1, jj]) { jj in range of [0, ii-1] }
         = max(f[k, ii-1], prices[ii] + max(f[k-1, jj] - prices[jj]))
f[0, ii] = 0; 0 times transation makes 0 profit
f[k, 0] = 0; if there is only one price data point you can't make any money no matter how many times you can trade
"""
class Solution_2:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0: return 0
        K = 2 # number of max transation allowed
        maxProf = 0
        f = [[0] * len(prices)] * (K + 1)
        for kk in range(1,K + 1):
            tmpMax = f[kk-1][0] - prices[0]
            for ii in range(1,len(prices)):
                f[kk][ii] = max(f[kk][ii-1], prices[ii] + tmpMax)
                tmpMax = max(tmpMax, f[kk-1][ii] - prices[ii])
                maxProf = max(f[kk][ii], maxProf)

        return maxProf

File: leetcode/C++/0123-h-best-time-to-buy-and-sell-stock-iii.cpp
// DP
// phase 1, 3, 5:
// f[i][j] = max(f[i - 1][j], f[i-1][j-1] + prices[i-1] - prices[i-2])
// phase 2, 4:
// f[i][j] = max(f[i-1][j-1], f[i-1][j] + prices[i-1] - prices[i-2])
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n == 0) return 0;
        
        vector<vector<int>> f(n + 1, vector<int>(5 + 1, 0));
        int i, j;
        f[0][1] = 0;
        for (i = 2; i <= 5; i++) 
            f[0][i] = INT_MIN;
        
        for (i = 1; i <= n; ++i) {
            for (j = 1; j<=5; j += 2) {
                f[i][j] = f[i-1][j];
                if (j > 1 && i >= 2 && f[i-1][j-1] != INT_MIN)
                    f[i][j] = max(f[i][j], f[i-1][j-1] + prices[i-1] - prices[i-2]);
            }
            for (j = 2; j<=4; j += 2) {
                f[i][j] = f[i - 1][j - 1];
                if (i >= 2 && f[i-1][j] != INT_MIN)
                    f[i][j] = max(f[i][j], f[i-1][j] + prices[i-1] - prices[i-2]);
            }
        }
        return max(f[n][1], max(f[n][3], f[n][5]));
    }
};